packages/wdio-cucumber-framework/src/index.ts

```diff
@@ -48,123 +59,106 @@ const {
         )
-        this._hasTests = true
-        this._cucumberFeaturesWithLineNumbers = this._config.cucumberFeaturesWithLineNumbers || []
-        this._eventBroadcaster = new EventEmitter()
-        this._eventDataCollector = new EventDataCollector(this._eventBroadcaster)
+
+        this._cucumberOpts.formatOptions = {
+            _reporter: this._reporter,
+            _cid: this._cid,
+            _specs: this._specs,
+            _eventEmitter: this._eventEmitter,
+            _scenarioLevelReporter: this._cucumberOpts.scenarioLevelReporter,
+            _tagsInTitle: this._cucumberOpts.tagsInTitle,
+            _ignoreUndefinedDefinitions: this._cucumberOpts.ignoreUndefinedDefinitions,
+            _failAmbiguousDefinitions: this._cucumberOpts.failAmbiguousDefinitions
+        }
 
         const builder = new Gherkin.AstBuilder(uuidFn)
-        const matcher = new Gherkin.GherkinClassicTokenMatcher(this._config.cucumberOpts?.featureDefaultLanguage || 'en')
+        const matcher = new Gherkin.GherkinClassicTokenMatcher(
+            this._cucumberOpts.language
+        )
         this.gherkinParser = new Gherkin.Parser(builder, matcher)
 
-        this._specs = this._specs.map((spec) => (
-            /**
-             * as Cucumber doesn't support file:// formats yet we have to
-             * remove it before adding it to Cucumber
-             */
-            spec.startsWith(FILE_PROTOCOL)
-                ? url.fileURLToPath(spec)
-                : spec
-        ))
-        const featurePathsToRun = this._cucumberFeaturesWithLineNumbers.length > 0
-            ? this._cucumberFeaturesWithLineNumbers
-            : this._specs
-
-        this._pickleFilter = new PickleFilter({
-            cwd: this._cwd,
-            featurePaths: featurePathsToRun,
-            names: this._cucumberOpts.names as string[],
-            tagExpression: this._cucumberOpts.tagExpression
-        })
+        /**
+         * as Cucumber doesn't support file:// formats yet we have to
+         * remove it before adding it to Cucumber
+         */
+        this._specs = this._specs.map((spec) =>
+            spec.startsWith(FILE_PROTOCOL) ? url.fileURLToPath(spec) : spec
+        )
 
-        const reporterOptions = {
-            capabilities: this._capabilities,
-            ignoreUndefinedDefinitions: Boolean(this._cucumberOpts.ignoreUndefinedDefinitions),
-            failAmbiguousDefinitions: Boolean(this._cucumberOpts.failAmbiguousDefinitions),
-            tagsInTitle: Boolean(this._cucumberOpts.tagsInTitle),
-            scenarioLevelReporter: Boolean(this._cucumberOpts.scenarioLevelReporter)
-        }
-        this._cucumberReporter = new CucumberReporter(this._eventBroadcaster, this._pickleFilter, reporterOptions, this._cid, this._specs, this._reporter)
+        // backwards compatibility for tagExpression usage
+        this._cucumberOpts.tags = this._cucumberOpts.tags ? this._cucumberOpts.tags : this._cucumberOpts.tagExpression
```

Comment:
```suggestion
        this._cucumberOpts.tags = this._cucumberOpts.tags || this._cucumberOpts.tagExpression
```

Reply:
Also, should we add a deprecation warning here?

Reply:
It's marked as deprecated in both the types and documentation. However, if necessary, we can generate a warning message as below. I would appreciate your thoughts on this approach.

```
if (this._cucumberOpts.tagExpression) {
    log.warn("'tagExpression' is deprecated. Use 'tags' instead.");
}
```

Reply:
Your suggestion is a good approach. I believe there are a lot of non TS users out there that won't see the type deprecation. While this log message probably will be hard to spot as well, it is the best we can do for now until we have a better deprecation strategy in general.